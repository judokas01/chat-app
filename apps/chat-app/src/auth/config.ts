export const AUTH_MODULE_SALT_ROUNDS = 10
