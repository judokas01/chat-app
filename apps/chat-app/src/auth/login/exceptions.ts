import { HttpException, HttpStatus } from '@nestjs/common'

export class UserNotFoundError extends HttpException {
    constructor() {
        super('User was not found.', HttpStatus.NOT_FOUND)
    }
}

export class InvalidPasswordError extends HttpException {
    constructor() {
        super('Invalid username and password combination.', HttpStatus.BAD_REQUEST)
    }
}

export class InvalidRenewTokenRequestError extends HttpException {
    constructor() {
        super('Invalid renew token request.', HttpStatus.BAD_REQUEST)
    }
}
