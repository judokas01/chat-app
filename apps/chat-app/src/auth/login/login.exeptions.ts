import { HttpException, HttpStatus } from '@nestjs/common'

export class UserNotFoundError extends HttpException {
    constructor() {
        super('User was not found.', HttpStatus.NOT_FOUND)
    }
}

export class InvalidPasswordError extends HttpException {
    constructor() {
        super('Invalid username and password combination.', HttpStatus.BAD_REQUEST)
    }
}
