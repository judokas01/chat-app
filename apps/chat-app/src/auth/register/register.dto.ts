import { IsEmail, IsNotEmpty, IsStrongPassword } from 'class-validator'

export class RegisterRequest {
    @IsEmail()
    email: string

    @IsStrongPassword({ minLength: 8, minSymbols: 0 })
    password: string

    @IsNotEmpty()
    userName: string
}
