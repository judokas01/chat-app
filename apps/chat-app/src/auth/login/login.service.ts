import { randomUUID } from 'crypto'
import { Inject, Injectable } from '@nestjs/common'
import { User, UserInput } from '@root/common/entities/user.entity'
import bcrypt from 'bcrypt'
import { JwtService } from '@nestjs/jwt'
import { IUserRepository } from '../repository/user-repository.interface'
import { AUTH_MODULE_SALT_ROUNDS } from '../config'
import {
    InvalidPasswordError,
    InvalidRenewTokenRequestError,
    UserNotFoundError,
} from './exceptions'

@Injectable()
export class LoginService {
    constructor(
        @Inject(IUserRepository) private prismaRepository: IUserRepository,
        private jwtService: JwtService,
    ) {}

    login = async ({
        password,
        userName,
    }: Pick<UserInput, 'userName' | 'password'>): Promise<LoginServiceResult> => {
        const user = await this.prismaRepository.findByUserName(userName)

        if (!user) {
            throw new UserNotFoundError()
        }

        const isPasswordValid = await bcrypt.compare(password, user.password)

        if (!isPasswordValid) {
            throw new InvalidPasswordError()
        }

        const [accessToken, renewToken] = await Promise.all([
            this.getJwtToken(user),
            this.createRenewToken(user.id),
        ])

        return {
            accessToken,
            renewToken,
        }
    }

    renewToken = async (args: RenewRequest): Promise<LoginServiceResult> => {
        const { authToken, renewToken } = args

        const authTokenPayload: JWTPayload = this.jwtService.decode(authToken)

        const user = await this.prismaRepository.findById(authTokenPayload.sub)

        const storedRenewToken = await this.prismaRepository.findRenewTokenByUserId(
            authTokenPayload.sub,
        )

        const isRenewTokenValid = await bcrypt.compare(renewToken, storedRenewToken ?? '')

        if (!user || !storedRenewToken || !isRenewTokenValid) {
            throw new InvalidRenewTokenRequestError()
        }

        const [newAccessToken, newRenewToken] = await Promise.all([
            this.getJwtToken(user),
            this.createRenewToken(user.id),
        ])

        return {
            accessToken: newAccessToken,
            renewToken: newRenewToken,
        }
    }

    private createRenewToken = async (userId: string): Promise<string> => {
        const token = randomUUID()

        const hashedToken = await bcrypt.hash(token, AUTH_MODULE_SALT_ROUNDS)

        await this.prismaRepository.updateRenewToken(userId, hashedToken)

        return token
    }

    private getJwtToken = async (user: User): Promise<string> => {
        const jwtPayload: JWTPayload = { sub: user.id, userName: user.userName }
        return await this.jwtService.signAsync(jwtPayload)
    }
}

export type LoginServiceResult = {
    accessToken: string
    renewToken: string
}

type RenewRequest = {
    authToken: string
    renewToken: string
}

type JWTPayload = { sub: string; userName: string }
