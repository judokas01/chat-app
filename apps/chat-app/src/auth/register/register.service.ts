import { Inject, Injectable } from '@nestjs/common'
import { User, UserInput } from '@root/common/entities/user.entity'
import * as bcrypt from 'bcrypt'
import { IUserRepository } from '../repository/user-repository.interface'
import { AUTH_MODULE_SALT_ROUNDS } from '../config'
import { UserAlreadyExistsError } from './exceptions'

@Injectable()
export class RegisterService {
    constructor(@Inject(IUserRepository) private prismaRepository: IUserRepository) {}

    async register(user: UserInput): Promise<User> {
        const found = await this.prismaRepository.findByUserName(user.userName)

        if (found) {
            throw new UserAlreadyExistsError()
        }

        return await this.prismaRepository.createOne({
            ...user,
            password: await this.hashPassword(user.password),
        })
    }

    private hashPassword = async (password: string): Promise<string> => {
        return await bcrypt.hash(password, AUTH_MODULE_SALT_ROUNDS)
    }
}
