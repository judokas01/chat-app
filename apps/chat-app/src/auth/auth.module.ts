import { Module } from '@nestjs/common'
import { LoginService } from './login/login.service'
import { LogoutService } from './logout/logout.service'
import { RegisterService } from './register/register.service'
import { AuthenticateService } from './authenticate/authenticate.service'
import { JWT } from './common/jwt.module'

@Module({
    imports: [JWT],
    providers: [LoginService, LogoutService, RegisterService, AuthenticateService],
})
export class AuthModule {}
