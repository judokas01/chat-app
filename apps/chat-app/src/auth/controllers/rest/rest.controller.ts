import { Controller, Post, Body } from '@nestjs/common'
import { LoginRequest, LoginServiceResult, RenewRequest } from '@root/auth/login/login.dto'
import { LoginService } from '@root/auth/login/login.service'
import { RegisterRequest, RegisteredUser } from '@root/auth/register/register.dto'
import { RegisterService } from '@root/auth/register/register.service'
import { toRegisterResponse } from './mappers'

@Controller('auth')
export class RestController {
    constructor(
        private registerService: RegisterService,
        private loginService: LoginService,
    ) {}

    @Post('/register')
    async register(@Body() registerReq: RegisterRequest): Promise<RegisteredUser> {
        const result = await this.registerService.register(registerReq)
        return toRegisterResponse(result)
    }

    @Post('/login')
    async login(@Body() loginReq: LoginRequest): Promise<LoginServiceResult> {
        return await this.loginService.login(loginReq)
    }

    @Post('/renew')
    async renewToken(@Body() renewReq: RenewRequest): Promise<LoginServiceResult> {
        return await this.loginService.renewToken(renewReq)
    }
}
