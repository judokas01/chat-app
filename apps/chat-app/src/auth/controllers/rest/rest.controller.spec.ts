import { RegisterService } from '@root/auth/register/register.service'
import { LoginService } from '@root/auth/login/login.service'
import {
    TestInterfaceModule,
    getTestModuleWithInterface,
} from '@root/common/test-utilities/test-app/interface'
import { AuthModule } from '@root/auth/auth.module'
import { RestController } from './rest.controller'

describe('Auth Rest Controller tests', () => {
    let testModule: TestInterfaceModule

    beforeAll(async () => {
        testModule = await getTestModuleWithInterface({ module: AuthModule })
    })

    afterAll(async () => {
        await testModule.destroy()
    })

    beforeEach(async () => {
        await testModule.cleanDb()
    })

    it('should register', async () => {
        const response = await testModule.request
            .post('/auth/register')
            .send({ email: 'mail@aa.aa', password: 'VeryStr@ngPass!123', userName: 'userName' })
        console.log({ response: response.body })
    })
})
