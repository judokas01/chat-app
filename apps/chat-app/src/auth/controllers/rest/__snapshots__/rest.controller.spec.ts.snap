// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Rest Controller tests register + login should return 400 when credentials are incorrect 1`] = `
{
  "message": "Invalid username and password combination.",
  "statusCode": 400,
}
`;

exports[`Auth Rest Controller tests register should return invalid request when email is not email 1`] = `
{
  "error": "Bad Request",
  "message": [
    "email must be an email",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Rest Controller tests register should return invalid request when multiple properties are wrong 1`] = `
{
  "error": "Bad Request",
  "message": [
    "email must be an email",
    "password is not strong enough",
    "userName should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Rest Controller tests register should return invalid request when password is week 1`] = `
{
  "error": "Bad Request",
  "message": [
    "password is not strong enough",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Rest Controller tests register should return invalid request when userName is undefined 1`] = `
{
  "error": "Bad Request",
  "message": [
    "userName should not be empty",
  ],
  "statusCode": 400,
}
`;
